### PR TITLE
Add encode and decode to Input and Output

### PR DESCRIPTION
**Pull request by zachgk**
This adds a full serialization rather than just part of content or properties to the Input and Output. It also requires adding equals and hash to them as well.

### CODE REVIEW

File: api/src/main/java/ai/djl/modality/Input.java [hunk 1]
```diff
@@ -328,6 +352,113 @@ public NDList getAsNDList(NDManager manager, int index) {
...
+                }
+                contentRead += newRead;
+            }
+            input.add(key, contents);
+        }
+    }
+
+    /** {@inheritDoc} * */
+    @Override
+    public boolean equals(Object o) {
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/modality/Input.java:444]: This is too heavy, we either use a id or just use Identity

File: api/src/main/java/ai/djl/modality/Output.java [hunk 2]
```diff
@@ -70,4 +79,69 @@ public String getMessage() {
...
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        if (!super.equals(o)) {
+            return false;
+        }
+        Output output = (Output) o;
+        return code == output.code && Objects.equals(message, output.message);
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/modality/Output.java:139]: We cannot just based on message and code

Comment (zachgk) [reply api/src/main/java/ai/djl/modality/Output.java:144]: It has a check based on `super.equals()` a few lines above. This is the IntelliJ equals template

File: api/src/main/java/ai/djl/modality/Input.java (hunk 1 above)
Comment (zachgk) [reply api/src/main/java/ai/djl/modality/Input.java]: I moved it to a separate `deepEquals()` method. So, we can keep the functionality but mostly use the lighter Object equals

**Review by frankfliu — APPROVED** [review]

**Pull request merged**